src/PrestaShopBundle/Security/Admin/UserProvider.php

```diff
@@ -0,0 +1,151 @@
+use Symfony\Component\Security\Core\Exception\UsernameNotFoundException;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+
+/**
+ * Responsible for retrieving Employee entities for the Symfony security components.
+ */
+final class UserProvider implements UserProviderInterface
+{
+    /**
+     * @var Context
+     */
+    private $legacyContext;
+
+    /**
+     * @var CacheItemPoolInterface
+     */
+    private $cache;
+
+    /**
+     * @var CommandBusInterface
+     */
+    private $queryBus;
+
+    /**
+     * @param LegacyContext $context
+     * @param CacheItemPoolInterface $cache
+     * @param CommandBusInterface $queryBus
+     */
+    public function __construct(
+        LegacyContext $context,
+        CacheItemPoolInterface $cache,
+        CommandBusInterface $queryBus
+    ) {
+        $this->legacyContext = $context->getContext();
+        $this->cache = $cache;
+        $this->queryBus = $queryBus;
+    }
+
+    /**
+     * Fetch the Employee entity that matches the given username.
+     * Cache system doesn't supports "@" character, so we rely on a sha1 expression.
+     *
+     * @param string $username
+     *
+     * @return Employee
+     *
+     * @throws InvalidArgumentException
+     * @throws UsernameNotFoundException
+     */
+    public function loadUserByUsername($username)
+    {
+        $cacheKey = sha1($username);
+        $cachedEmployee = $this->cache->getItem("app.employees_${cacheKey}");
+
+        if ($cachedEmployee->isHit()) {
+            return $cachedEmployee->get();
+        }
+
+        try {
+            /** @var EmployeeForAuthentication $employeeForAuthentication */
+            $employeeForAuthentication = $this->queryBus->handle(
+                GetEmployeeForAuthentication::fromEmail($username)
+            );
+            $employee = new Employee(
+                new LegacyEmployee($employeeForAuthentication->getEmployeeId()->getValue())
+            );
+            $employee->setRoles($employeeForAuthentication->getRoles());
+
+            $cachedEmployee->set($employee);
+            $this->cache->save($cachedEmployee);
+
+            return $cachedEmployee->get();
+        } catch (AuthenticatingEmployeeNotFoundException $e) {
+            throw new UsernameNotFoundException(
+                sprintf('Username "%s" does not exist.', $username),
+                0,
+                $e
+            );
+        }
+    }
+
+    /**
+     * Reload an Employee and returns a fresh instance.
+     *
+     * @param UserInterface $employee
+     *
+     * @return Employee
+     */
+    public function refreshUser(UserInterface $employee)
+    {
+        if (!$employee instanceof Employee) {
+            throw new UnsupportedUserException(
+                sprintf('Instances of "%s" are not supported.', get_class($employee))
+            );
+        }
+
+        return $this->loadUserByUsername($employee->getUsername());
```

Comment:
Shouldn't you use the email since it seems to be the reference for the authentication system?

Reply:
Email is the username. Because of `\Symfony\Component\Security\Core\User\UserInterface` I must use `getUsername()` method

Reply:
We can add comments to make it very clear, but indeed because of Symfony forcing us to use `loadUserByUsername()` we'll have to deal with this ambiguity 😞

Reply:
@jolelievre `$this->loadUserByUsername($employee->getEmail());` cannot be done because `getEmail()` method does not exist in Symfony's `UserInterface`, I'd have to break interface to do that

Reply:
I've added some comments to explain it in the code